File: converted/source2/source/zombie/fileSystem/DiskFileDevice$DiskFile.java
package zombie.fileSystem;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.RandomAccessFile;
import zombie.core.logger.ExceptionLogger;
import zombie.fileSystem.DiskFileDevice.1;

final class DiskFileDevice$DiskFile implements IFile {
   final DiskFileDevice m_device;
   RandomAccessFile m_file;
   InputStream m_inputStream;
   final IFile m_fallthrough;
   boolean m_use_fallthrough;

   DiskFileDevice$DiskFile(IFile var1, DiskFileDevice var2) {
      this.m_device = var2;
      this.m_fallthrough = var1;
      this.m_use_fallthrough = false;
   }

   public boolean open(String var1, int var2) {
      File var3 = new File(var1);
      boolean var4 = (var2 & 1) != 0;
      if (var4 && !var3.exists() && this.m_fallthrough != null) {
         this.m_use_fallthrough = true;
         return this.m_fallthrough.open(var1, var2);
      } else {
         try {
            if ((var2 & 16) == 0) {
               this.m_file = new RandomAccessFile(var1, FileOpenMode.toStringMode(var2));
            } else {
               this.m_inputStream = new FileInputStream(var1);
            }

            return true;
         } catch (IOException var6) {
            ExceptionLogger.logException(var6);
            return false;
         }
      }
   }

   public void close() {
      if (this.m_fallthrough != null) {
         this.m_fallthrough.close();
      }

      if (this.m_file != null || this.m_inputStream != null) {
         try {
            if (this.m_file != null) {
               this.m_file.close();
            }

            if (this.m_inputStream != null) {
               this.m_inputStream.close();
            }
         } catch (IOException var2) {
            ExceptionLogger.logException(var2);
         }

         this.m_file = null;
         this.m_inputStream = null;
         this.m_use_fallthrough = false;
      }
   }

   public boolean read(byte[] var1, long var2) {
      if (this.m_use_fallthrough) {
         return this.m_fallthrough.read(var1, var2);
      } else if (this.m_file == null) {
         return false;
      } else {
         try {
            return this.m_file.read(var1, 0, (int)var2) == var2;
         } catch (IOException var5) {
            ExceptionLogger.logException(var5);
            return false;
         }
      }
   }

   public boolean write(byte[] var1, long var2) {
      if (this.m_use_fallthrough) {
         return this.m_fallthrough.write(var1, var2);
      } else if (this.m_file == null) {
         return false;
      } else {
         try {
            this.m_file.write(var1, 0, (int)var2);
            return true;
         } catch (IOException var5) {
            ExceptionLogger.logException(var5);
            return false;
         }
      }
   }

   public byte[] getBuffer() {
      return this.m_use_fallthrough ? this.m_fallthrough.getBuffer() : null;
   }

   public long size() {
      if (this.m_use_fallthrough) {
         return this.m_fallthrough.size();
      } else if (this.m_file == null) {
         return 0L;
      } else {
         try {
            return this.m_file.length();
         } catch (IOException var2) {
            ExceptionLogger.logException(var2);
            return 0L;
         }
      }
   }

   // $VF: Unable to simplify switch on enum
   // Please report this to the Vineflower issue tracker, at https://github.com/Vineflower/vineflower/issues with a copy of the class file (if you have the rights to distribute it!)
   public boolean seek(FileSeekMode var1, long var2) {
      if (this.m_use_fallthrough) {
         return this.m_fallthrough.seek(var1, var2);
      } else if (this.m_file == null) {
         return false;
      } else {
         try {
            this.m_file.seek(switch (1.$SwitchMap$zombie$fileSystem$FileSeekMode[var1.ordinal()]) {
               case 2 -> this.m_file.getFilePointer();
               case 3 -> this.m_file.length();
            });
            return true;
         } catch (IOException var5) {
            ExceptionLogger.logException(var5);
            return false;
         }
      }
   }

   public long pos() {
      if (this.m_use_fallthrough) {
         return this.m_fallthrough.pos();
      } else if (this.m_file == null) {
         return 0L;
      } else {
         try {
            return this.m_file.getFilePointer();
         } catch (IOException var2) {
            ExceptionLogger.logException(var2);
            return 0L;
         }
      }
   }

   public InputStream getInputStream() {
      return this.m_inputStream;
   }

   public IFileDevice getDevice() {
      return this.m_device;
   }

   public void release() {
      this.getDevice().destroyFile(this);
   }
}

File: ProjectZomboidSwitch/src/zombie/fileSystem/DiskFileDevice$DiskFile.cpp
#include <string>
#include "zombie/fileSystem/DiskFileDevice$DiskFile.h"

namespace zombie {
namespace fileSystem {

bool DiskFileDevice$DiskFile::open(const std::string& var1, int var2) {
    // Check if READ flag is set and file does not exist, use fallthrough if available
    bool readFlag = (var2 & 1) != 0;
    std::ifstream testFile(var1);
    bool exists = testFile.good();
    testFile.close();
    if (readFlag && !exists && m_fallthrough) {
        m_use_fallthrough = true;
        return m_fallthrough->open(var1, var2);
    } else {
        try {
            if ((var2 & 16) == 0) {
                // Not a stream: open for read/write
                m_file = std::make_shared<RandomAccessFile>(var1, FileOpenMode::toStringMode(var2));
            } else {
                // Stream mode
                m_inputStream = std::make_shared<FileInputStream>(var1);
            }
            return true;
        } catch (const std::exception& ex) {
            ExceptionLogger::logException(ex);
            return false;
        }
    }
}

void DiskFileDevice$DiskFile::close() {
    // Close fallthrough if used
    if (m_fallthrough) {
        m_fallthrough->close();
    }

    // Close file and input stream if they exist
    try {
        if (m_file) {
            m_file->close();
        }
        if (m_inputStream) {
            m_inputStream->close();
        }
    } catch (const std::exception& ex) {
        ExceptionLogger::logException(ex);
    }

    m_file.reset();
    m_inputStream.reset();
    m_use_fallthrough = false;
}

bool DiskFileDevice$DiskFile::read(byte[] var1, long var2) {
    // If using fallthrough, delegate read
    if (m_use_fallthrough && m_fallthrough) {
        return m_fallthrough->read(var1, var2);
    }

    try {
        if (m_file) {
            // RandomAccessFile read
            return m_file->read(var1, var2);
        } else if (m_inputStream) {
            // FileInputStream read
            return m_inputStream->read(var1, var2);
        }
    } catch (const std::exception& ex) {
        ExceptionLogger::logException(ex);
    }
    return false;
}

bool DiskFileDevice$DiskFile::write(byte[] var1, long var2) {
    // If using fallthrough, delegate write
    if (m_use_fallthrough && m_fallthrough) {
        return m_fallthrough->write(var1, var2);
    }

    try {
        if (m_file) {
            // RandomAccessFile write
            return m_file->write(var1, var2);
        }
    } catch (const std::exception& ex) {
        ExceptionLogger::logException(ex);
    }
    return false;
}

long DiskFileDevice$DiskFile::size() {
    // If using fallthrough, delegate size
    if (m_use_fallthrough && m_fallthrough) {
        return m_fallthrough->size();
    }
    try {
        if (m_file) {
            return m_file->size();
        }
    } catch (const std::exception& ex) {
        ExceptionLogger::logException(ex);
    }
    return 0;
}

bool DiskFileDevice$DiskFile::seek(FileSeekMode var1, long var2) {
    // If using fallthrough, delegate seek
    if (m_use_fallthrough && m_fallthrough) {
        return m_fallthrough->seek(var1, var2);
    }
    try {
        if (m_file) {
            return m_file->seek(var1, var2);
        }
    } catch (const std::exception& ex) {
        ExceptionLogger::logException(ex);
    }
    return false;
}

long DiskFileDevice$DiskFile::pos() {
    // If using fallthrough, delegate pos
    if (m_use_fallthrough && m_fallthrough) {
        return m_fallthrough->pos();
    }
    try {
        if (m_file) {
            return m_file->pos();
        }
    } catch (const std::exception& ex) {
        ExceptionLogger::logException(ex);
    }
    return 0;
}

InputStream DiskFileDevice$DiskFile::getInputStream() {
    // If using fallthrough, delegate getInputStream
    if (m_use_fallthrough && m_fallthrough) {
        return m_fallthrough->getInputStream();
    }
    if (m_inputStream) {
        return m_inputStream.get();
    }
    return nullptr;
}

IFileDevice DiskFileDevice$DiskFile::getDevice() {
    // If using fallthrough, delegate getDevice
    if (m_use_fallthrough && m_fallthrough) {
        return m_fallthrough->getDevice();
    }
    return this;
}

void DiskFileDevice$DiskFile::release() {
    // If using fallthrough, delegate release
    if (m_use_fallthrough && m_fallthrough) {
        m_fallthrough->release();
        return;
    }
    // Release file and input stream
    m_file.reset();
    m_inputStream.reset();
    m_use_fallthrough = false;
}

} // namespace fileSystem
} // namespace zombie
